src/sample/trainers/forward_dynamics.py

```diff
@@ -0,0 +1,378 @@
+
+        for _ in range(self.max_epochs):
+            batch: tuple[Tensor, Tensor, Tensor]
+            for batch in dataloader:
+                observations, actions, hiddens = batch
+                observations = observations.to(device)
+                actions = actions.to(device)
+                obs_imaginations, hiddens = (
+                    observations[:, : -self.imagination_length],
+                    hiddens[:, 0].to(device),
+                )
+
+                self.optimizers[OPTIMIZER_NAME].zero_grad()
+
+                obses_next_hat_dist: Distribution
+                loss_imaginations: list[Tensor] = []
+                for i in range(self.imagination_length):
+                    action_imaginations = actions[
+                        :, i : -self.imagination_length + i
+                    ]  # a_i:i+T-H, (B, T-H, *)
+                    obs_targets = observations[
+                        :,
+                        i + 1 : observations.size(1) - self.imagination_length + i + 1,
+                    ]  # o_i+1:T-H+i+1, (B, T-H, *)
+                    if i > 0:
+                        action_imaginations = action_imaginations.flatten(
+                            0, 1
+                        )  # (B', *)
+                        obs_targets = obs_targets.flatten(0, 1)  # (B', *)
+
+                    obses_next_hat_dist, next_hiddens = self.forward_dynamics.model(
+                        obs_imaginations, action_imaginations, hiddens
+                    )
+                    loss = -obses_next_hat_dist.log_prob(obs_targets).mean()
+                    loss_imaginations.append(loss)
+                    obs_imaginations = obses_next_hat_dist.rsample()
+
+                    if i == 0:
+                        obs_imaginations = obs_imaginations.flatten(
+                            0, 1
+                        )  # (B, T-H, *) -> (B', *)
+                        hiddens = next_hiddens.movedim(2, 1).flatten(
+                            0, 1
+                        )  # h'_i, (B, D, T-H, *) -> (B, T-H, D, *) -> (B', D, *)
+
+                loss = self.imagination_average_method(torch.stack(loss_imaginations))
+                loss.backward()
+
+                prefix = "forward-dynamics/"
+                metrics_imagination = {
+                    f"{prefix}/loss/imagination_{i}": loss_item.item()
+                    for i, loss_item in enumerate(loss_imaginations)
+                }
+                grad_norm = torch.cat(
+                    [
+                        p.grad.flatten()
+                        for p in self.forward_dynamics.model.parameters()
+                        if p.grad is not None
+                    ]
+                ).norm()
+                metrics_grad_norm = {f"{prefix}/grad_norm": grad_norm.item()}
+                mlflow.log_metrics(metrics_imagination, step=self.global_step)
+                mlflow.log_metrics(metrics_grad_norm, step=self.global_step)
```

Comment:
`log_metrics`は一度で複数のログを行えます。あとはいくつかログの値を修正しました。
```suggestion
                loss = self.imagination_average_method(torch.stack(loss_imaginations))
                loss.backward()

                metrics = {"loss/average": loss.item()}
                for i, loss_item in enumerate(loss_imaginations, start=1):
                    metrics[f"loss/imagination/{i}"] = loss_item.item()
                
                metrics["grad norm"] = torch.cat(
                    [
                        p.grad.flatten()
                        for p in self.forward_dynamics.model.parameters()
                        if p.grad is not None
                    ]
                ).norm()
                
                mlflow.log_metrics({f"forward-dynamics/{k}": v for k, v in metrics.items()}, self.global_step)
```

Reply:
https://github.com/MLShukai/pamiq-vrchat/pull/45/commits/c705f0e1146e968cc45ff8f6ec0b6fe57f4a9e3a
修正しました